### DISCUSSION TITLE
FL32 file format

### DISCUSSION THREAD

**Discussion by RichardKirk** [ImageMagick/ImageMagick#2847] [Development]
I have used the .fl32 for nearly 20 years. It is supported in Baselight. I have written a reader and writer for Octave. You are unlikely to see it anywhere else.

It is really simple. The header is 4 32-bit integers: magic, ny, nx, nc, followed by the image data as a packed float[ny][nx][nc]. That is the whole of the  fl32 format document.

I could try adding a reader and writer if this is straightforward, and no-one objects.

**Comment by dlemstra** [comment]
Do you have a link to some documentation about that format? I cannot seem to find any information about this image format.

**Comment by RichardKirk** [comment]
> On 19 Nov 2020, at 16:55, Dirk Lemstra <notifications@github.com> wrote:
> Do you have a link to some documentation about that format? I cannot seem to find any information about this image format.
> 
No, there isn’t any. It is not much used outside FilmLight R&D. But it is really useful for saving float data without having to mess with all of the other stuff to save as EXR or 32-bit TIFF. There was something on Wotsit once, but I can’t find Wotsit, let alone my bit.

The magic number is 842222662, which should be ‘FL32’ in ASCII on a big-endian machine. This is the only bit of documentation I forgot. There is no explicit limit on the size or the number of channels. I have used 2-channel images, but this is rare. 1,3, and 4 channels are used by some others. I usually use it with conventional XY image axes but there is no standard.

If you need it, I can wrote a formal fl32 format document. Or I can supply test images.

Cheers
Richard Kirk

**Reply by fmw42** [reply]
You can also use NetPBM PFM for a simple float data format. See http://netpbm.sourceforge.net/doc/pfm.html

**Comment by RichardKirk** [comment]
> On 19 Nov 2020, at 17:36, Fred Weinhaus <notifications@github.com> wrote:
> 
> You can also use NetPBM PFM for a simple float data format. See http://netpbm.sourceforge.net/doc/pfm.html <http://netpbm.sourceforge.net/doc/pfm.html>
Yes, I could. I had forgot I tried that (or something like it). It was good having the initial string readable, but it might give problems if the closing ‘/n’ character did not leave you at a 4-byte boundary. I can’t remember whether Windows was still using "/r/n” or whatever, but it was still worrying. The fl32 header is 16 bytes so you are always on a boundary.

**Comment by urban-warrior** [comment]
> If you need it, I can wrote a formal fl32 format document. Or I can supply test images.

Please do.  We'll take a look @ adding support to ImageMagick.

**Comment by RichardKirk** [comment]
FL32 image format.

This image format has been used within FilmLight Ltd. It is the simplest
floating-point format we could come up with. The aim has been to make the
format as simple as possible, and to avoid the use of any format-dependent
tags for orientation, colour space, pixel size, and the like.

No rights are reserved and no attribution is requested.

Richard Kirk
FilmLight
Sat 21 Nov 2020


* File name

The extensions .fl32 and .FL32 are used for this format.


* Header

The header of the image is 4 32-bit signed integer little-endian values.
In 'C++' notation this might be " int head[4]; "

The values are...
head[0] = 'magic number' must be 842222662
head[1] = lines per image
head[2] = pixels per line
head[3] = values per pixel

The 'magic number' corresponds to the bytes "FL32" on a little-endian
system. It will read as something else on a big-endian system. We have
not supported a big-endian system, but this should not be too hard.

The 'lines', 'pixels', and 'values' integers cannot be negative or zero.
This range is reserved, but be have no plans to use them.

There are no limits on the positive range for these values. A typical
reader might be expected to support 1, 3, or 4 values per pixel. These
are conventionally taken as being...
1 - monochrome
3 - red, green, blue
4 - red, gren, blue, alpha

There are no restrictions on the use and naming of these values.
We have used CIE L*a*b*, CMYK and othr spaces. We could use many more than
4 channels for hyperspectral images.

We usually have pixels running left to right as along the x-axis, and
lines running from bottom to top as along the y-axis. But the format
implies no particular orientation.

* Data

The header is followed immediately by the data as a packed float array.
The data array starts at byte 16.

In 'C++' notation this might be " float [lines][pixels][values]; "


* Example header dump...

An integer header dump might look like this...
> od -i face.fl32 | head
0000000         842222662             240             166               4
0000020        1051109031      1051240617      1053609165      1065353216
0000040        1050977445      1051240617      1053345993      1065353216

The image is in portrat mode, with 240 lines, 166 pixels per line,
and 4 values per pixel.

> od -a face.fl32 | head
0000000    F   L   3   2   ? nul nul nul   ? nul nul nul eot nul nul nul
0000020    ?   ?   ?   >   ?   ?   ?   >   ?   ?   ?   > nul nul  80   ?
0000040    ?   ?   ?   >   ?   ?   ?   >   ?   ?   ?   > nul nul  80   ?

An ASCII dump shows the characters in the 'magic number'...

> od -f face.fl32 | head                   
0000000     1.043651e-08    3.363116e-43    2.326155e-43    5.605194e-45
0000020     3.254902e-01    3.294118e-01    4.000000e-01    1.000000e+00
0000040     3.215686e-01    3.294118e-01    3.921569e-01    1.000000e+00

A float dump shows the image values...

> od -f face.fl32 | head                   
0000000     1.043651e-08    3.363116e-43    2.326155e-43    5.605194e-45
0000020     3.254902e-01    3.294118e-01    4.000000e-01    1.000000e+00
0000040     3.215686e-01    3.294118e-01    3.921569e-01    1.000000e+00

These are RGB+alpha values. The RGB white point is usually taken to be
1.0 but HDR images can go beyond this. There is no assumption of the
display space, so this may be linear or with gamma, or other tone curve.

In this example we have a 4th alpha channel which is 1.0. This makes for
neat columns, but RGB images are also widely used. If you want RGB, you
can read this and discard the 4th channel. If you have RGB and you want to
write a 4-channel image, you can add a 4th channel packed with 1.0. These
are useful options to have in a reader, but the FL32 standard does not
offer any opinion on whether this is the right thing to do.


* C++ example image read as packed pixels

  FILE* in = fopenError(name, "r");
  if (in == NULL)
      return 1;       // Cannot open image
      
  int head[4];
  if (4 != fread(head, sizeof(int), 4, in) ||
      head[0] != 842222662 ||
      head[1] < 1 || head[2] < 1 || head[3] < 1 ||
      head[3] > 4)    // arbitrary limit, more is legal
   } return 2;        // Cannot read header

   int size = head[1]*head[2]*head[3];
   float* data = new float[size];
   if (size != fread(data, sizeof(float), size, in))
       return 3;      // Cannot read data


* Octave / Matlab planar image read in planar format

function dfrm = loadFloat32(fname)

% image = loadFloat32(fname) loads a 32-bit float file
% as float[line][pixel][channel] with the native byte ordering
% The header is 4 32-bit ints written with the same ordering.
% The first number should be "FL32" on a big-endian machine.

fid = fopen(fname, 'r');
if (fid == -1)
  error('cannot open %s', fname);
end


% FL32 header
H = fread(fid, 4, 'int32');
if H(1) != 842222662
  error('%s is not a float32 image', fname);
end

channels = H(2);
pixels   = H(3);
lines    = H(4);

[A count] = fread(fid, lines*pixels*channels, 'float32');
if count < lines*pixels*channels
  error('%s file truncated', fname);
end

fclose(fid);
dfrm = permute(reshape(A, [lines pixels channels]), [3 2 1]);




Quick fl32 format document. Hope it will do.




Example RGBa image. Should look right in typical display RGB. Let me know if you want something different.

Thanks.
Richard Kirk

**Comment by fmw42** [comment]
I am a bit confused. You start by saying that the data is only non-negative integers.

"The 'lines', 'pixels', and 'values' integers cannot be negative or zero."

But then you show example values such as 1.043651e-08, which is a fraction less than 1.

Please clarify.

What application use do you normally use for this format?

Without float values, it is limited for use with HDRI. (I note that PFM and TIFF allow fractional values and negative values). So without negative values, it would be limiting for some digital terrain data (below sea level needs negative values). So they would be clipped.

I am not saying not to implement this. But it would be nice to understand better it use case.

**Comment by urban-warrior** [comment]
> Example RGBa image ...

Where?  Do you have a link to an example RGBa FL32 image we can download?

**Comment by RichardKirk** [comment]
> On 21 Nov 2020, at 18:55, Fred Weinhaus <notifications@github.com> wrote:
> 
> 
> I am a bit confused. You start by saying that the data is only non-negative integers.
> 
> "The 'lines', 'pixels', and 'values' integers cannot be negative or zero."
> 
> 

Apologies if it isn’t clear.

The three integer values in the header that give the number of lines per image, pixels per line, and values per pixel cannot be zero or negative. I could have used unsigned values, but this would give an array that wold be so enormous that it is not practical. This does not set any limit on the float values themselves.
> But then you show example values such as 1.043651e-08, which is a fraction less than 1.
> 
There is no restriction on the float values in the image. L*a*b* has negative chroma values. Even spaces such as CIE XYZ may have to support a small amount of ’negative light’ if only so that local averages are not biased towards the positive by clipping at zero. There is nothing forbidding NaN values, though it is usually much more sensible to pick these up as close to their origin, and to to pack them into an image to be discovered later.
> What application use do you normally use for this format?
> 
The reads and writes are fast. Baselight supports these but they are not generally used in the ordinary motion picture pipeline. However, I can output as these, and then read them into ‘octave’ or Matlab. The reads are significantly faster than most of the more general library read and write routines. I can also read them into several other image processing programs including some of the utility programs paced with Baselight.

In general, I find the FL32 format is most useful when we want to grab the image values without worrying about being truncated to a shorter format or losing precision, and then reading to some other program. The images are not compact, so the format is most useful for transferring data short distances between programs, rather than sending from one site to another.

I have used 2-channel images for flow maps - (x,y) offsets between two images. This is pretty rare but it has got used. Again, these were signed offsets.

We could have made a FL16 format that used half floats. Half floats are a good match for the actual ranges in real scene-based linear RGB. However, we have to use a library routine to convert between the 32- and 16-bit formats. They ought to be completely defined but sometime there are differences in how the values are rounded or the unnormalised values are handled. I would argue that the FL32 format is for people who would rather save everything as 32-bit floats than worry about the processing to save to a smaller format, and the possible effect on the values.
 
> Without float values, it is limited for use with HDRI. Without negative values, it would be limiting for some digital terrain data (below sea level needs negative values). So they would be clipped.

So, yes - the pixel data is pull 32-bit floats with all values allowed.

Cheers
Richard Kirk

**Reply by fmw42** [reply]
Thanks. Sorry for my confusion. That makes much more sense now if pixel data is full float.

**Comment by RichardKirk** [comment]
> On 21 Nov 2020, at 19:53, Cristy <notifications@github.com> wrote:
> 
> 
> Example RGBa image ...
> 
> Where? Do you have a link to an example RGBa FL32 image we can download?
> 
> 
Face.fl32, attached to the message.

If it hasn’t made it through to the thread, I can send it somewhere else. It is the one that I used then describing the header dumps in the document.

Cheers
Richard Kirk

**Comment by RichardKirk** [comment]
> On 21 Nov 2020, at 18:55, Fred Weinhaus <notifications@github.com> wrote:
> 
> 
> I am a bit confused. You start by saying that the data is only non-negative integers.
> 
> "The 'lines', 'pixels', and 'values' integers cannot be negative or zero."
> 
> But then you show example values such as 1.043651e-08, which is a fraction less than 1.
> 
Ah. I see where this has come from. The magic number is “FL32” when dumb as 4 bytes, and 842222662 when dumped as an integer. These same bytes will give 1.043651e-08 (a meaningless number) when they are interpreted a 32-bit floats by the 'od’ tool. But the 4 header fields are always 32-bit integers.

Cheers
Richard Kirk

**Comment by urban-warrior** [comment]
Are there any of these images in the wild?  Supporting the format would not require much work, but at the same time if only you and two other people in the world are using the format, there is less incentive to add support for FL32 to ImageMagick.

Also, we don't see any image attachments.  Post a link to a FL32 image that we can download from a website.

**Comment by RichardKirk** [comment]
> On 21 Nov 2020, at 20:37, Cristy <notifications@github.com> wrote:
> Are there any of these images in the wild? Supporting the format would not require much work, but at the same time if only you and two other people in the world are using the format, there is less incentive to add support for FL32 to ImageMagic.
> 
You are right. I understand this. There are a few places which have output this format form a scanner so we could read it, but most of our work is with people who protect their images with vigour - imagine the stink if we let slip a frame from a Harry Potter, or Star Wars film (and we have handled both). The telephone was not a useful invention until the invention of the other telephone. I have no-one who wants to exchange floating-point images in the format because they have no tools that support it, but they can pack them into OpenEXR or TIFF, and usually at a pinch this is what they do. But it would be better if I could get a straight image of their data, and not to have to wonder whether the NaN values are just outside the half-float range, or whether there are no negative values because something inside the tool has clipped them.

It is clearly up to you guys what you support, and I won’t sulk if you don’t. But I think this format has its uses.
i
> 
Here we go…

File face.fl32 has been published
The URL for downloading is  https://private.filmlight.ltd.uk/e082a5bec6857cac/face.fl32 <https://private.filmlight.ltd.uk/e082a5bec6857cac/face.fl32>
It will be removed from the server in 7 days.

Cheers
Richard Kirk

**Comment by urban-warrior** [comment]
Look for FL32 support in the next release of IMv7 within a day or two.  By default, 4 channels are interpreted as RGBA.  If its CMYK, simply add `-set colorspace CMYK` to your command line.

**Comment by RichardKirk** [comment]
> On 22 Nov 2020, at 13:25, Cristy <notifications@github.com> wrote:
> 
> Look for FL32 support in the next release of IMv7 within a day or two.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/ImageMagick/ImageMagick/discussions/2847#discussioncomment-135727>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAMCRAPNYW65XGXWSU3TOTDSREGNTANCNFSM4T3UWFRA>.
> 
Whoo! Thanks a lot. I have been using gnu parallel all this weekend with a little program of mine, and I reckon that would have saved a significant part of it.

Cheers
Richard Kirk

**Comment by RichardKirk** [comment]
Hi.

I have been busy the last week on other stuff. I have just refreshed imagemagick using ‘homebrew’ on my Big Sur Mac, and it works fine with 3- and 4-channel .fl32 images. Wonderful!

Thanks again.
Richard Kirk
